waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -386,22 +387,53 @@ when defined(rln) or (not defined(rln) and not defined(rlnzerokit)):
 
     return proofBytes
 
-  proc proofVerify*(rlnInstance: RLN[Bn256], data: openArray[byte],
-    proof: RateLimitProof): bool =
+  proc getMerkleRoot*(rlnInstance: RLN[Bn256]): MerkleNodeResult =
+    # read the Merkle Tree root after insertion
+    var
+      root {.noinit.}: Buffer = Buffer()
+      rootPtr = addr(root)
+      getRootSuccessful = get_root(rlnInstance, rootPtr)
+    if (not getRootSuccessful): return err("could not get the root")
+    if (not (root.len == 32)): return err("wrong output size")
+
+    var rootValue = cast[ptr MerkleNode] (root.`ptr`)[]
+    return ok(rootValue)
+
+  proc validateRoot*(rlnInstance: RLN[Bn256], proof: RateLimitProof): BoolResult =
+    # Validate against the local merkle tree
+    let localTreeRoot = rlnInstance.getMerkleRoot()
+    if not localTreeRoot.isOk():
+      return err(localTreeRoot.error())
+    if localTreeRoot.value() == proof.merkleRoot:
+      return ok(true)
+    else:
+      trace "Local tree root does not match the root sent by peer", localTreeRoot=localTreeRoot, messageRoot=proof.merkleRoot
+      return ok(false)
+
+  proc proofVerify*(rlnInstance: RLN[Bn256], data: openArray[byte], proof: RateLimitProof): BoolResult =
     var
       proofBytes = serialize(proof, data)
       proofBuffer = proofBytes.toBuffer()
       f = 0.uint32
     trace "serialized proof", proof = proofBytes.toHex()
 
+    let rootIsValid = validateRoot(rlnInstance, proof)
+
+    if rootIsValid.isErr():
+      return err(rootIsValid.error())
```

Comment:
Please note that this could have been

```nim
let rootIsValid = ?validateRoot(rlnInstance, proof)
```

However, to enable future refactoring, for tighter error enums, it has been handled manually